main/src/library/Iterator.flix

```diff
@@ -20,6 +20,66 @@ pub enum Iterator[a] {
 
 namespace Iterator {
 
+    ///
+    /// Returns `Some(x)` if `x` is the first element of `iter`.
+    ///
+    /// Returns `None` if `iter` is empty.
+    ///
+    /// Consumes the first element of the iterator.
+    ///
+    pub def head(iter: Iterator[a]): Option[a] & Impure = {
+        let Iterator(done, next) = iter;
+        if (done())
+            None
+        else
+            Some(next())
+    }
+
+    ///
+    /// Returns an iterator of all integers between `b` (inclusive) and `e` (exclusive).
+    ///
+    /// Returns an empty iterator if `b >= e`.
+    ///
+    pub def range(b: Int32, e: Int32): Iterator[Int32] & Impure = List.range(b, e) |> toIter
```

Comment:
This constructs the entire list and then makes it lazy. That works, but is not so great if the range is longer.

Instead, you should use a counter and construct two functions `done` and `next`. (See the documentation on how to use references).

Reply:
I see, thanks.
My intention was to make a "dumb" implementation first, make the tests pass and then redo it.